servant-client/test/Servant/ClientTestUtils.hs

```diff
@@ -145,6 +145,24 @@ instance ToDeepQuery Filter where
     , (["name"], Just (Text.pack name'))
     ]
 
+-- MultiVerb test endpoint
+
+data ErrorResponse a
+
+data UserNotFound
+
+type instance ResponseType (ErrorResponse a) = a
+
+type GetUserVerb =
+  Capture "personId" Int
+  :> MultiVerb
+    'GET
+    '[JSON]
+    '[ ErrorResponse UserNotFound,
+       Respond 200 "User found" Person
+     ]
+    (Maybe Person)
```

Comment:
I am looking at https://github.com/wireapp/wire-server/blob/0e57e6be7eccaf23f4b69569855d8893e33a5c18/libs/wire-api/src/Wire/API/User.hs#L783-L822

If I understand well: 

* I have to define the errors for the endpoint ([`data RegisterError`](https://github.com/wireapp/wire-server/blob/0e57e6be7eccaf23f4b69569855d8893e33a5c18/libs/wire-api/src/Wire/API/User.hs#L783))
* I have to define another error (?) as defined in [Error/Brig.hs](https://github.com/wireapp/wire-server/blob/0e57e6be7eccaf23f4b69569855d8893e33a5c18/libs/wire-api/src/Wire/API/Error/Brig.hs#L47)
* Then a type-level list of `ErrorResponse`s parametrised with some of the errors from Brig (`ErrorResponse 'AllowlistError`).
* And then derive AsUnion with `  deriving (AsUnion RegisterErrorResponses) via GenericAsUnion RegisterErrorResponses RegisterError` which ties everything together. 

Which part of it is redundant outside of wire-server?

Reply:
I think you can ignore the error stuff, it's kinda neat, but not required and not really finished.  From my understanding the only thing you need is the `AsUnion` instance that explains how you can compute a `Union '[a]` from a `ResponseType a` (following the naming in the definition of `MultiVerb`.

`ResponseType a` can map to a union type, then the `AsUnion` instance is trivial or not necessary, not sure how that looks.

If it maps to an ordinary type, you can map different values of that type to different response types of your union, thus allowing servant to do it's magic.

Not sure that answers your question, or whether I know what I'm talking about...  :)

Do you have any particular commit and corresponding error message that you are struggling with?

Reply:
Alright, so I've got this: https://github.com/haskell-servant/servant/pull/1766/files#diff-a4411c940d784b1fef13144b967568c54a5e3986b2c3f5449c5c4297af69a94fR148-R174

However now I have this error:

```
test/Servant/ClientTestUtils.hs:164:13: error: [GHC-83865]
    • Couldn't match type ‘()’ with ‘Person’
      Expected: Maybe Person -> Union (ResponseTypes res)
        Actual: Maybe Person -> Union ('[()] .++ '[()])
    • In the expression: maybeToUnion (toUnion @GetUserMultiVerbErrors)
      In an equation for ‘toUnion’:
          toUnion = maybeToUnion (toUnion @GetUserMultiVerbErrors)
      In the instance declaration for ‘AsUnion res (Maybe Person)’
    |
164 |   toUnion = maybeToUnion (toUnion @GetUserMultiVerbErrors)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I am not sure where the need for `()` is coming from, since in the Wire codebase it does not appear.

Reply:
hum, that smells like somewhere the wrong instance of `AsUnion` or `AsConstructor` is picked? have you tried `-ddump-{ds,tc,simpl,splices,...}`?

Reply:
Problem has disappeared with your suggestion below. :) I even added an example of possible error in the cookbook and the documentation of `AsUnion`. Please check them out!